TestHostNetworkPod flake

https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48175/unit-tests-arm64_istio/1732642864627716096

cc @howardjohn Looks like related to https://github.com/istio/istio/pull/47893 


I got the following clue but not sure exactly why.

Add logs:
```
diff --git a/pkg/test/util/assert/tracker.go b/pkg/test/util/assert/tracker.go
index 2e46157b6a..e337948120 100644
--- a/pkg/test/util/assert/tracker.go
+++ b/pkg/test/util/assert/tracker.go
@@ -38,7 +38,9 @@ func NewTracker[T comparable](t test.Failer) *Tracker[T] {
 func (t *Tracker[T]) Record(event T) {
        t.mu.Lock()
        defer t.mu.Unlock()
+       t.t.Logf("====== Record func: append before %+v, append %+v", t.events, event)
        t.events = append(t.events, event)
+       t.t.Logf("====== Record func: append after %+v", t.events)
 }

 // Empty asserts the tracker is empty
@@ -65,7 +67,9 @@ func (t *Tracker[T]) WaitOrdered(events ...T) {
                                t.t.Fatalf("got events %v, want %v", t.events, event)
                        }
                        // clear the event
+                       t.t.Logf("------ WaitOrdered func: clear before %+v, clear %+v", t.events, event)
                        t.events = t.events[1:]
+                       t.t.Logf("------ WaitOrdered func: clear after %+v", t.events)
                        return nil
                }, retry.Timeout(time.Second))
        }
```



Got the following logs:
```
=== RUN   TestHostNetworkPod
2023-12-07T16:35:26.091992Z	info	cluster "fake" kube client started
2023-12-07T16:35:26.094278Z	info	controllers	starting	controller=crd watcher
2023-12-07T16:35:26.094368Z	info	kube	kube controller for  synced after 12.591µs
    tracker.go:41: ====== Record func: append before [], append pod1/add
    tracker.go:43: ====== Record func: append after [pod1/add]
    tracker.go:41: ====== Record func: append before [pod1/add], append pod1/update
    tracker.go:43: ====== Record func: append after [pod1/add pod1/update]
    tracker.go:41: ====== Record func: append before [pod1/add pod1/update], append pod2/add
    tracker.go:43: ====== Record func: append after [pod1/add pod1/update pod2/add]
    tracker.go:70: ------ WaitOrdered func: clear before [pod1/add pod1/update pod2/add], clear pod1/add
    tracker.go:72: ------ WaitOrdered func: clear after [pod1/update pod2/add]
    tracker.go:70: ------ WaitOrdered func: clear before [pod1/update pod2/add], clear pod1/update
    tracker.go:72: ------ WaitOrdered func: clear after [pod2/add]
    tracker.go:70: ------ WaitOrdered func: clear before [pod2/add], clear pod2/add  <-- pod2 add here 
    tracker.go:72: ------ WaitOrdered func: clear after []
    tracker.go:41: ====== Record func: append before [], append pod1/delete
    tracker.go:43: ====== Record func: append after [pod1/delete]
    tracker.go:41: ====== Record func: append before [pod1/delete], append pod2/delete  <-- pod2 delete here, delete event before update.
    tracker.go:43: ====== Record func: append after [pod1/delete pod2/delete]
    tracker.go:67: got events [pod1/delete pod2/delete], want pod2/update    <-- check pod2 update fail
2023-12-07T16:35:26.118408Z	info	kube	Controller terminated
2023-12-07T16:35:26.118410Z	info	controllers	stopped	controller=crd watcher
--- FAIL: TestHostNetworkPod (0.03s)
```

--- a/pilot/pkg/serviceregistry/kube/controller/pod_test.go
+++ b/pilot/pkg/serviceregistry/kube/controller/pod_test.go
@@ -161,8 +161,9 @@ func TestHostNetworkPod(t *testing.T) {
 
 	createPod("128.0.0.1", "pod1")
 	assert.Equal(t, c.pods.getPodKeys("128.0.0.1"), []types.NamespacedName{{Name: "pod1", Namespace: "ns"}})
-
+	events.WaitOrdered("pod1/add", "pod1/update")
 	createPod("128.0.0.1", "pod2")
+	events.WaitOrdered("pod2/add", "pod2/update")
 	assert.Equal(t, sets.New(c.pods.getPodKeys("128.0.0.1")...), sets.New(
 		types.NamespacedName{Name: "pod1", Namespace: "ns"},
 		types.NamespacedName{Name: "pod2", Namespace: "ns"},
@@ -174,14 +175,7 @@ func TestHostNetworkPod(t *testing.T) {
 	}
 	pods.Delete("pod1", "ns")
 	pods.Delete("pod2", "ns")
-	events.WaitOrdered(
-		"pod1/add",
-		"pod1/update",
-		"pod2/add",
-		"pod2/update",
-		"pod1/delete",
-		"pod2/delete",
-	)
+	events.WaitOrdered("pod1/delete", "pod2/delete")
 }
 
 // Regression test for https://github.com/istio/istio/issues/20676

--- a/pkg/test/util/assert/tracker.go
+++ b/pkg/test/util/assert/tracker.go
@@ -19,6 +19,7 @@ import (
 	"sync"
 	"time"
 
+	"istio.io/istio/pkg/ptr"
 	"istio.io/istio/pkg/test"
 	"istio.io/istio/pkg/test/util/retry"
 )
@@ -55,6 +56,7 @@ func (t *Tracker[T]) Empty() {
 func (t *Tracker[T]) WaitOrdered(events ...T) {
 	t.t.Helper()
 	for _, event := range events {
+		var err error
 		retry.UntilSuccessOrFail(t.t, func() error {
 			t.mu.Lock()
 			defer t.mu.Unlock()
@@ -62,12 +64,47 @@ func (t *Tracker[T]) WaitOrdered(events ...T) {
 				return fmt.Errorf("no events")
 			}
 			if t.events[0] != event {
-				t.t.Fatalf("got events %v, want %v", t.events, event)
+				// Exit early instead of continuing to retry
+				err = fmt.Errorf("got events %v, want %v", t.events, event)
+				return nil
 			}
 			// clear the event
 			t.events = t.events[1:]
 			return nil
 		}, retry.Timeout(time.Second))
+		if err != nil {
+			t.t.Fatal(err)
+		}
 	}
+	t.Empty()
+}
+
+// WaitUnordered waits for an event to happen, in any order
+func (t *Tracker[T]) WaitUnordered(events ...T) {
+	t.t.Helper()
+	want := map[T]struct{}{}
+	for _, e := range events {
+		want[e] = struct{}{}
+	}
+	retry.UntilSuccessOrFail(t.t, func() error {
+		t.mu.Lock()
+		defer t.mu.Unlock()
+		if len(t.events) == 0 {
+			return fmt.Errorf("no events (want %v)", want)
+		}
+		got := t.events[0]
+		if _, f := want[got]; !f {
+			return fmt.Errorf("got events %v, want %v", t.events, want)
+		}
+		// clear the event
+		t.events[0] = ptr.Empty[T]()
+		t.events = t.events[1:]
+		delete(want, got)
+
+		if len(want) > 0 {
+			return fmt.Errorf("still waiting for %v", want)
+		}
+		return nil
+	}, retry.Timeout(time.Second), retry.BackoffDelay(time.Millisecond))
 	t.Empty()
 }

